### PR TITLE
[BugFix] Fix the bug when calculating weight_comm_time in SPMDSolver

### PR DESCRIPTION
**Pull request by V1ewLine**
Because of a bug in the code, the weight_comm_time was always set to zero. As a result, when choosing the best pipeline plan, the weight_comm_time was not included in the overall calculation of the total time, leading to the selection of an incorrect pipeline scheme.

### CODE REVIEW

**Comment by zyeric** [comment]
thanks for your contribution! seems it is a bug.
Can you add a test to check the communication time for weight update is correct after your modification?
You can check the example at https://github.com/microsoft/nnscaler/blob/main/tests/autodist/spmd_solver/test_follow.py#L174.
The test may be a simple nn.Module with one `torch.nn.Linear` in it, and check the communication cost after calculating the cost vector.
Thanks again.

**Comment by V1ewLine** [comment]
> thanks for your contribution! seems it is a bug. Can you add a test to check the communication time for weight update is correct after your modification? You can check the example at https://github.com/microsoft/nnscaler/blob/main/tests/autodist/spmd_solver/test_follow.py#L174. The test may be a simple nn.Module with one `torch.nn.Linear` in it, and check the communication cost after calculating the cost vector. Thanks again.

Okay, I will upload the file after the test succeed.

**Comment by V1ewLine** [comment]
> 果您同意 CLA

@microsoft-github-policy-service agree

> @V1ewLine please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Comment by V1ewLine** [comment]
@microsoft-github-policy-service agree

File: tests/autodist/spmd_solver/test_weight_time.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+import os
+from pathlib import Path
+from nnscaler.cli.trainer_args import TrainerArgs
+from nnscaler.graph.parser.converter import to_fx_graph, to_ir_graph
+from nnscaler.autodist.model_graph import ModelGraph
+from nnscaler.autodist.autodist_config import AutoDistConfig
+from nnscaler.autodist.spmd_solver import SPMDSolver
+from nnscaler.parallel import ComputeConfig
+
+class Attention(torch.nn.Module):
```
Comment (zyeric) [inline tests/autodist/spmd_solver/test_weight_time.py:19]: If the module is identical with `test_follow` and you want to reuse it, you can `from tests.autodist.spmd_solver.test_follow import AttentionModel`

File: tests/autodist/spmd_solver/test_attention_follow.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 - allowed_partition_dims:
   - 1,0
-  - 1,1
+  - 0,0
```
Comment (zyeric) [inline tests/autodist/spmd_solver/test_attention_follow.yaml:3]: Why we need to update this? You can add a new yaml file and pass it to the test file if you only want to partition along batch dim.

File: tests/autodist/spmd_solver/test_weight_time.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        mesh_desc=cfg.mesh_desc,
+        autodist_config=cfg,
+        stage_num=1,
+        micro_batch_num=cfg.update_freq,
+    )
+
+    is_correct = False
+    for p in spmd_solver.partition_info:
+        for i in p:
+            if(i.weight_update_time != 0):
```
Comment (zyeric) [inline tests/autodist/spmd_solver/test_weight_time.py:100]: in python we don't need `()` for `if`.
please check more carefully, seems in this model, only `linear` node's weight update time > 0, other nodes do not have weight tensor.

**Review by zyeric — CHANGES_REQUESTED** [review]

File: tests/autodist/spmd_solver/test_weight_time.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+- allowed_partition_dims:
+  - 1,0
```
Comment (zyeric) [inline tests/autodist/spmd_solver/test_weight_time.yaml:2]: you can remove L2, L9 and L15 to enforce autodist to generate a data parallel plan

File: tests/autodist/spmd_solver/test_weight_time.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import os
+from pathlib import Path
+from nnscaler.graph.parser.converter import to_fx_graph, to_ir_graph
+from nnscaler.autodist.model_graph import ModelGraph
+from nnscaler.autodist.autodist_config import AutoDistConfig
+from nnscaler.autodist.spmd_solver import SPMDSolver
+from tests.autodist.spmd_solver.test_follow import AttentionModel
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason='CUDA unavailable')
+def test_follow_AttentionModel():
```
Comment (zyeric) [inline tests/autodist/spmd_solver/test_weight_time.py:16]: rename to `test_weight_comm_time`

File: tests/autodist/spmd_solver/test_weight_time.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    #double check
+    is_correct = True
+    for i in range(spmd_solver.graph.op_num):
+        for j in range(spmd_solver.get_op_partition_count(i)):        
+            cost_desc = spmd_solver.calc_partition_cost(i, j)
+            if not cost_desc.weight_update_time==cost_info[i][j].weight_update_time:
+                is_correct = False
+
+    assert is_correct == True
```
Comment (zyeric) [inline tests/autodist/spmd_solver/test_weight_time.py:60]:
```python
cur_op = spmd_solver.get_operator(i)
if 'linear' in cur_op.op_name:
    # check there exist at least one partition whose `weight_update_time` > 0
else:
    # assert all `weight_update_time` = 0
```

**Review by zyeric — CHANGES_REQUESTED** [review]

File: tests/autodist/spmd_solver/__pycache__/__init__.cpython-310.pyc
Comment (zyeric) [inline tests/autodist/spmd_solver/__pycache__/__init__.cpython-310.pyc:1]: please remove these pyc files, thanks

File: tests/autodist/spmd_solver/test_weight_comm_time.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                for j in range(spmd_solver.get_op_partition_count(i))):
+                continue
+            else:
+                is_correct = False
+                break
+        else:
+            if not all(cost_info[i][j].weight_update_time == 0 
+                    for j in range(spmd_solver.get_op_partition_count(i))):
+                is_correct = False
+                break
```
Comment (zyeric) [inline tests/autodist/spmd_solver/test_weight_comm_time.py:67]:
```python
weight_comm_times = [desc.weight_update_time for desc in cost_info[i]]
if 'linear' in cur_op.op_name:
    assert any(comm > 0 for comm in weight_comm_times)
else:
    assert all(comm == 0 for comm in weight_comm_times)
```

File: tests/autodist/spmd_solver/test_weight_comm_time.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import os
+from pathlib import Path
+from nnscaler.graph.parser.converter import to_fx_graph, to_ir_graph
+from nnscaler.autodist.model_graph import ModelGraph
+from nnscaler.autodist.autodist_config import AutoDistConfig
+from nnscaler.autodist.spmd_solver import SPMDSolver
+from tests.autodist.spmd_solver.test_follow import AttentionModel
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason='CUDA unavailable')
+def test_follow_AttentionModel():
```
Comment (zyeric) [inline tests/autodist/spmd_solver/test_weight_comm_time.py:16]: remember to refine the test name

**Review by zyeric — APPROVED** [review]

**Review by 0xWJ — APPROVED** [review]
